src/Microsoft.Framework.Runtime/ApplicationHostContext.cs

```diff
@@ -65,34 +76,30 @@ public class ApplicationHostContext
             {
                 var lockFileReader = new LockFileReader();
                 var lockFile = lockFileReader.Read(projectLockJsonPath);
-                validLockFile = lockFile.IsValidForProject(Project);
 
-                if (validLockFile || skipLockFileValidation)
+                string message;
+                validLockFile = lockFile.IsValidForProject(Project, out message);
+
+                if (skipLockFileValidation || validLockFile)
                 {
                     NuGetDependencyProvider.ApplyLockFile(lockFile);
-
-                    DependencyWalker = new DependencyWalker(new IDependencyProvider[] {
-                        ProjectDepencyProvider,
-                        NuGetDependencyProvider,
-                        referenceAssemblyDependencyResolver,
-                        gacDependencyResolver,
-                        unresolvedDependencyProvider
-                    });
+                }
+                else
+                {
+                    // Don't add NuGetDependencyProvider to DependencyWalker
+                    dependencyProviders = dependencyProviders.Where(d => d != NuGetDependencyProvider);
+                    _diagnostics.Add(new LockFileValidationMessage(message, projectLockJsonPath));
                 }
             }
-
-            if ((!validLockFile && !skipLockFileValidation) || !lockFileExists)
+            else
             {
-                // We don't add NuGetDependencyProvider to DependencyWalker
-                // It will leave all NuGet packages unresolved and give error message asking users to run "dnu restore"
-                DependencyWalker = new DependencyWalker(new IDependencyProvider[] {
-                    ProjectDepencyProvider,
-                    referenceAssemblyDependencyResolver,
-                    gacDependencyResolver,
-                    unresolvedDependencyProvider
-                });
+                // Don't add NuGetDependencyProvider to DependencyWalker
+                dependencyProviders = dependencyProviders.Where(d => d != NuGetDependencyProvider);
```

Comment:
I just checked and it does work but I would like to understand how... You assign a `WhereIterator` to a reference that is a source for this same iterator. Probably Where captures the original reference somewhere but it is not obvious.